README.md


		 WebTerm  - Web Terminal 



 DESCRIPTION
 ==========================================================================

  Wterm is terminal which make be able to communicate with unix shell via 
 HTTP. it could be used over firewall that allow only HTTP (port 80).
 there are three modules in this package. wtcgi.cgi, wterm, wtserver.
 "wterm" is a client interface like CRT. "wtcgi.cgi" is traditional 
 cgi program comunicates server with unix domain socket. and "wtserver" 
 is a domain socket server which manages pseudo terminal of sub process 
 "/bin/sh"

 The latest information about Wterm can be found at :
 http://underground.or.kr/project/webterm/


 INSTALL & USAGE 
 ==========================================================================

 First, you should move your tarball to cgi-bin directory could be excutable 
 and read-writable by web user in your web server. 
 ( bcoz, server creates unix domain socket for comunication with wtcgi.cgi ) 
 
 And next, extract tar file  and type 'make'
 if there were no bugs in procedure and three modules which refered above 
 created, then server installation completly done.  
 if not, plz config your detail set in each source file or send email to me.
 
 Next, run server with privilege  equal to 
 web server
 ex)
	./wtserver &

 and move wterm into your linux box which use as terminal client 
 Now, you can connect your server with your wterm
 ex)
	./wterm -h 192.168.1.200 -p 80 -l /cgi-bin/wtcgi.cgi


 CHANGES
 ==========================================================================

 Confer 'ChangeLog' in this package


 COPYING
 ==========================================================================

 This program is free software; you can  redistribute it and/or modify
 it under the terms  of the GNU General Public  License as published by
 the  Free Software Foundation; either version  2,  or (at your option)
 any later version.

 This program is distributed  in the hope that  it will be  useful, but
 WITHOUT    ANY  WARRANTY;  without   even   the  implied  warranty  of
 MERCHANTABILITY  or  FITNESS FOR  A PARTICULAR   PURPOSE.  See the GNU
 General Public License for more details.

 You should  have received a  copy of  the GNU  General Public  License
 along with   this  program; if   not,  write  to   the Free   Software
 Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.


 WARNNING !  It's for only educational purpose and warning to ppls
 about senstive security of MSN messenger.
 i have no responsibility for usnig this program for another purpose.

 WTERM - Web Terminal
 (C) COPYRIGHT Seunghyun Seo
 All right reserverd ( Aug 27, 2002 )


 CONTACT
 ======================================================================

 Send bug (error) reports, questions and comments to below email

 Seunghyun Seo, seo@igrus.inha.ac.kr or frog@hackerslab.com
 http://igrus.inha.ac.kr/~seo/






ChangeLog

ver 0.0.2:
	expanded buf size 1024 to 8192
	added setitimer to set timer , it replaces original alram call.


ver 0.0.1:
	first released, only works on linux


kbhit.c
#include <termios.h>
#include <term.h>
#include <unistd.h>

static struct termios initial_settings, new_settings;
static int peek_character = -1;

void init_keyboard();
void close_keyboard();
int kbhit();
int readch();

void init_keyboard()
{
	tcgetattr(0,&initial_settings);
	new_settings = initial_settings;
	new_settings.c_lflag &= ~ICANON;
	new_settings.c_lflag &= ~ECHO;
	new_settings.c_lflag &= ~ISIG;
	new_settings.c_cc[VMIN] = 1;
	new_settings.c_cc[VTIME] = 0;
	tcsetattr(0, TCSANOW, &new_settings);
}

void close_keyboard()
{
	tcsetattr(0, TCSANOW, &initial_settings);
}

int kbhit()
{
	char ch;
	int nread;

	if(peek_character != -1)
		return 1;

	new_settings.c_cc[VMIN]=0;
	tcsetattr(0, TCSANOW, &new_settings);
	nread = read(0,&ch,1);
	new_settings.c_cc[VMIN]=1;
	tcsetattr(0, TCSANOW, &new_settings);

	if(nread == 1) {
		peek_character = ch;
		return 1;
	}
	return 0;
}

int readch()
{
	char ch;
	if(peek_character != -1) {
		ch = peek_character;
		peek_character = -1;
		return ch;
	}
	read(0,&ch,1);
	return ch;
}


wtcgi.c
/* wtcgi.c
 * truefinder, seo@igrus.inha.ac.kr
 */
#include <stdio.h>
#include <sys/socket.h>
#include <sys/un.h>
#include <signal.h>
#include <fcntl.h>
#include <stdlib.h>

#define TMPFILE	"/tmp/local.sock"

void print_header(void);

main()
{
	struct sockaddr_un to;
	int  s;
	int  tolen, len, n, maxlen;
	char msg[512], inbuf[512];
	FILE *fp;
	int i;
	char *ch;
	maxlen = 512;

	len = atol(getenv("CONTENT_LENGTH"));
	ch = (char*)malloc( sizeof(char)*(len+1) );
	fgets( ch , len + 1 , stdin);
	
	print_header();

#ifdef	DEBUGZ
	printf("CLIENT KEY IS LIKE  %c(%x)\n", *ch , *ch);
#endif


	s = socket(AF_UNIX,SOCK_STREAM,0);

	to.sun_family = AF_UNIX;       
	strcpy(to.sun_path, TMPFILE );
	tolen = sizeof(to.sun_family) + sizeof(to.sun_path);
	connect(s,(struct sockaddr *) &to, tolen);

	write( s, ch, 1 );
	while( (i =read(s,inbuf, sizeof(inbuf)))){
		printf("%s", inbuf);
		bzero( inbuf, sizeof(inbuf));
	}

	close(s);                      
}

void print_header(void)
{
	printf("Content-Type : text/html\n\n");
}



ptypair.c
#include <errno.h>
#include <fcntl.h>
#include <grp.h>
#include <stdlib.h>
#include <string.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <unistd.h>


/* get_master_pty() takes a double-indirect character pointer in which
 * to put a slave name, and returns an integer file descriptor.
 * If it returns < 0, an error has occurred.
 * Otherwise, it has returned the master pty file descriptor, and fills
 * in *name with the name of the corresponding slave pty.
 * Once the slave pty has been opened, you are responsible to free *name.
 */

int get_master_pty(char **name) { 
   int i, j;
   /* default to returning error */
   int master = -1;

   /* create a dummy name to fill in */
   *name = strdup("/dev/ptyXX");

   /* search for an unused pty */
   for (i=0; i<16 && master <= 0; i++) {
      for (j=0; j<16 && master <= 0; j++) {
         (*name)[8] = "pqrstuvwxyzPQRST"[i];
         (*name)[9] = "0123456789abcdef"[j];
         /* open the master pty */
         if ((master = open(*name, O_RDWR )) < 0) { 
            if (errno == ENOENT) {
               /* we are out of pty devices */
               free (*name);
               return (master);
            }
         }
      }
   }
   if ((master < 0) && (i == 16) && (j == 16)) {
      /* must have tried every pty unsuccessfully */
      free (*name);
      return (master);
   }

   /* By substituting a letter, we change the master pty
    * name into the slave pty name.
    */
   (*name)[5] = 't';

   return (master);
}

/* get_slave_pty() returns an integer file descriptor.
 * If it returns < 0, an error has occurred.
 * Otherwise, it has returned the slave file descriptor.
 */

int get_slave_pty(char *name) 
{ 

   struct group *gptr;
   gid_t gid;
   int slave = -1;

   /* chown/chmod the corresponding pty, if possible.
    * This will only work if the process has root permissions.
    * Alternatively, write and exec a small setuid program that
    * does just this.
    */
   if ((gptr = getgrnam("tty")) != 0) {
      gid = gptr->gr_gid;
   } else {
      /* if the tty group does not exist, don't change the
       * group on the slave pty, only the owner
       */
      gid = -1;
   }

   /* Note that we do not check for errors here.  If this is code
    * where these actions are critical, check for errors!
    */
   //chown(name, getuid(), gid);
   /* This code only makes the slave read/writeable for the user.
    * If this is for an interactive shell that will want to
    * receive "write" and "wall" messages, OR S_IWGRP into the
    * second argument below.
    */
   //chmod(name, S_IRUSR|S_IWUSR|S_IROTH| S_IWOTH);

   /* open the corresponding slave pty */
   slave = open(name, O_RDWR);

   return (slave);
}



wtserver.c
/* wtserver.c
 * truefinder seo@igrus.inha.ac.kr
 */
#include <errno.h>
#include <sys/ioctl.h>
#include <termios.h>
#include <unistd.h>
#include <string.h>
#include <stdlib.h>
#include <fcntl.h>
#include <sys/stat.h>
#include <sys/time.h>
#include <sys/socket.h>
#include <sys/un.h>
#include <assert.h>
#include <stdio.h>
#include <setjmp.h>
#include <signal.h>
#include "ptypair.h"

#define TMPFILE	"/tmp/local.sock"
#define BUFSIZE 4096

jmp_buf env ;
sigjmp_buf sigenv;
static int sfd , cfd ;


void 
close_domain(void)
{
#ifdef	DEBUGZ
	printf("DEBUG : server closed\n");
#endif
	unlink(TMPFILE);
}

void 
sig_handler( int signum )
{
	if ( signum == SIGALRM ) {
#ifdef	DEBUGZ
		printf("DEBUG : alarm\n");
#endif
		close(cfd) ;
		siglongjmp( sigenv, 1 );
	}
	
	if ( signum == SIGINT ) {
		exit(0);
	}
}

int
main( void )
{
	int master ; int pid ;
	char *name ; int i ; 
	char ch = 0;
	char buf[BUFSIZE];
	struct termios ot , t ;
	fd_set ready;
	int done = 0;
	int pipe;
	FILE *fp1, *fp2 ;
	int flag = 1;

	//socket
	struct sockaddr_un sa, ca;
	int  slen, clen ;

	//signal
	sigset_t sigs;
	struct sigaction sa_old;
	struct sigaction sa_new ;
	int mask;

	//timer
	struct sigaction siga;
	struct itimerval timer;


	master = get_master_pty(&name);
	if ( master < 0 ) {
		perror("ptypair: master open failed");
		exit(1);
	}

	pid = fork() ;
	if ( pid == 0 ) { //child
		int slave ;
		close(master);

		slave = get_slave_pty(name);
		if ( slave < 0 ) {
			perror("ptypair: slave open failed");
			exit(1);
		}
		free(name);

		setsid();

		ioctl(slave, TIOCSCTTY, NULL );

		dup2(slave, STDIN_FILENO );
		dup2(slave, STDOUT_FILENO );
		dup2(slave, STDERR_FILENO );
		
		execl("/bin/sh", "sh", 0 );

		exit(1);
	}

	//parent
	
	unlink(TMPFILE);
	atexit( close_domain);
	free(name);
	fp1 = fdopen( master, "r");	

	/*
	tcgetattr( master, &ot );
        t = ot ;
        t.c_lflag &= ~(ICANON | ISIG | ECHO | ECHOCTL | ECHOE |
                        ECHOK | ECHOKE | ECHONL | ECHOPRT );

        t.c_iflag |= IGNBRK;
        t.c_cc[VMIN] = 1;
        t.c_cc[VTIME] = 0;
        tcsetattr( master, TCSANOW, &t);
	*/

	timer.it_value.tv_sec = 0;
	timer.it_value.tv_usec = 250000;

	timer.it_interval.tv_sec = 0;
	timer.it_interval.tv_usec = 250000;

	signal( SIGALRM, sig_handler );
	signal( SIGINT, sig_handler );
	signal( SIGPIPE, SIG_IGN);

	sfd = socket( AF_UNIX, SOCK_STREAM, 0 );
	if ( sfd < 0 ) { perror("socket"); exit(-1); }

	sa.sun_family = AF_UNIX;
	strcpy( sa.sun_path, TMPFILE);
	slen = sizeof( sa.sun_family) + sizeof(sa.sun_path);

	bind( sfd, (struct sockaddr *) &sa, slen );
	listen( sfd, 5 );
	do {

		sigsetjmp( sigenv, 1 );

		cfd =  accept( sfd, (struct sockaddr*) &sa, &slen );
#ifdef DEBUGZ
		pritnf("DEBUG : accept client");
#endif

		read( cfd, &ch , 1 );
		write( master, &ch, 1 );

		//alarm(1); 
		setitimer( ITIMER_REAL, &timer, NULL );

		while(1){
			bzero( buf, sizeof(buf));
			i = read( master , buf, sizeof(buf)) ;
			if ( i <= 0   ) { 
				close( cfd ); break;
#ifdef DEBUGZ
				if ( i == 0 ) printf("it's i = 0 \n");
				else printf("it's i < 0\n");
#endif
			}
			write(cfd, buf, i);
#ifdef DEBUGZ
			printf("DEBUG : buf = [%s]", buf );
#endif
		}

		close( cfd ); 
		printf("DEBUG : sucked miss operation\n");

	} while (!done);

	/* this really doesn't matter because each time a master pty is
	** opened, the corresponding slave pty has its termios settings
	** reset
	**/
	//tcsetattr(STDIN_FILENO, TCSANOW, &ot);
	unlink(TMPFILE);
	exit(0);
}



wterm.c
/* wterm.c
 * truefinder, seo@igrus.inha.ac.kr
 */
#include <stdio.h>
#include <string.h>
#include <sys/types.h>
#include <time.h>
#include <sys/socket.h>
#include <unistd.h>
#include <signal.h>
#include <stdlib.h>
#include <netinet/in.h>
#include <fcntl.h>
#include <getopt.h>
#include "kbhit.h"

void print_banner(void);
void print_usage( char *argv[]);

#define BUFSIZE	4096
#define HTTP_METHOD "POST"
#define HTTP_PROTO "HTTP/1.0\r\nContent-Length: 1\r\n\r\n"

int
main( int argc, char *argv[] )
{
	char cmd[1024];
	char *msg_h = HTTP_METHOD; 	
	char *msg_t = HTTP_PROTO;
	char ch ;
	int len, i, j ;

	struct sockaddr_in s ;
	int sock; int n;
	char buf[BUFSIZE];
	char *needle;

	static char *optstr = "h:p:l:";
	char oc;
	char *host;
	char *location;
	char *port;

	fd_set fds ;

	if ( argc < 3 ) {
		print_usage(argv);
		exit(0);
	}

	while( (oc = getopt( argc, argv, optstr)) != -1 )  {
		switch (oc) {
			case 'h':
				host = optarg;
				break;
			case 'p':
				port = optarg;
				break;
			case 'l':
				location= optarg;
				break;
			default:
				exit(-1);
		}
	}
	printf("%s %s %s \n", host ,port, location);

	print_banner();


	init_keyboard();
	
	do {

		if ( kbhit()) {
			ch = readch();
			if (ch == '\n') ch = '\r';
			
		} else continue;

		sprintf( cmd , "%s %s %s%c", msg_h, location, msg_t, ch);
		len = strlen(cmd);
		
		sock = socket( PF_INET, SOCK_STREAM, 0 );
		if ( sock < 0 ){
			perror("socket");
			exit(0);
		}

		bzero( &s, sizeof(s) );
		s.sin_family = PF_INET ;
		s.sin_port = htons(atoi(port));
		s.sin_addr.s_addr = inet_addr(host);

		if ( connect( sock , (struct sockaddr*)&s, sizeof(s) ) <0 ) {
			perror("connect");
			exit(0);
		}

		write( sock, cmd , len );

		bzero( buf, sizeof(buf) );
		i = read( sock, buf, sizeof(buf) );

		needle = strstr(buf, "\r\n\r\n" );
		needle += 4;
		j = strlen( needle);
	
		write( STDOUT_FILENO, needle, j );
		
		close(sock);

	} while(1);

	close_keyboard();

	// Author : Seunghyun Seo 
	// Last update : at Aug 24 03:20:31 KST 2002
	// seo@igrus.inha.ac.kr , frog@hackerslab.com
				
}


void
print_usage(char* argv[])
{
	fprintf(stderr, "Usage : %s -h [ip] -p [port] -l [location]\n", argv[0] );
	fprintf(stderr, "Options:\n");
	fprintf(stderr, "\t[ip] host ip\n" );
	fprintf(stderr, "\t[port] port of webserver\n" );
	fprintf(stderr, "\t[location] location of wtcgi.cgi\n\n" );
	fprintf(stderr, "\tex)\n" );
	fprintf(stderr, "\t%s -h 192.168.1.1 -p 80 -l /cgi-bin/wtcgi.cgi\n", argv[0] );
}

void 
print_banner(void)
{
	printf("+-----------------------------------------------------------+\n");
	printf("|                                                           |\n");
	printf("|           Welcome to Web Terminal                         |\n");
	printf("|                                                           |\n");
	printf("|                            by truefinder                  |\n");
	printf("|                  seo@igrus.inha.ac.kr, frog@hackerslab.com|\n");
	printf("|                                                           |\n");
	printf("+-----------------------------------------------------------+\n");
}

